### ISSUE TITLE
Expand window properties exposed through rules

### ISSUE THREAD

**Issue by koekeishiya** [koekeishiya/kwm#284]
Improvements for #268 

Feel free to comment on things that could be useful to have, although be aware that they might not be possible to implement.

Additions:
 - (??) ~~Allow regexp for matching titles~~ Implemented.
 - (??) ~~kwmc add-role~~ Implemented.
 - ~~(??) override gap / padding~~ Not gonna be implemented.
 - ~~(??) callbacks~~ On hold..
 - ~~Assign window to space~~ Implemented.
 - ~~Automatically add to scratchpad~~ Implemented.
 - (??) Set size and position of floating windows ( width / height / maximize)

Pattern matching uses `std::regex` and `std::regex_match`.
For regular expression syntax, refer to ECMAscript specification or POSIX for detailed information.

**Comment by chrishoage** [comment]
> Allow regexp for matching titles

:+1: 

> kwmc add-role

:+1: 

> override gap / padding

Is this on a per container basis? or does it override the space's gap / padding setting? 

> callbacks

I still vote shell scripts.

**Comment by matthewdias** [comment]
priority order for tiling upon tree restore

**Comment by dominiklohmann** [comment]
For callbacks, I'd like to see callbacks for focus changes, window creation and so on. Do you have a list of things that you could (and want to) notify about?

**Comment by mickaelperrin** [comment]
If regexp support takes too much time to implement, maybe a much simpler approach would be helpful in lot of cases by adding an additional `name-match-mode` with the following options:
- `exact` the window title should match exactly the name
- `contains` the window title should contains the name pattern
- `starts-with` the window title should starts with the name pattern
- `ends-with` the window title should ends with the name pattern

**Comment by koekeishiya** [comment]
The next change to this system is most likely `space="id"`. This will allow you to assign a window to a specific space. Because space ids are numbered from `1 .. N` per display, this requires the `display` property to also be specified.

**Comment by mickaelperrin** [comment]
Additional ideas for floating windows:
- set size ( width / height / maximize)
- set position

But the rule must be applied only once to allow manual moving and changing size. That can be really helpful when you work on an IDE and you want to maximize some subwindows (commit / diff...) automatically.

**Comment by koekeishiya** [comment]
> The next change to this system is most likely space="id". This will allow you to assign a window to a specific space. Because space ids are numbered from 1 .. N per display, this requires the display property to also be specified (or otherwise default to the primary display - 0).

This has been implemented on master.
Example rules:

Assign itunes to space 1 of display 1.
`kwmc rule owner="iTunes" properties={space="1"; display="1"}`

Assign itunes to space 1 of the primary display (display 0).
`kwmc rule owner="iTunes" properties={space="1"}`

Note: Windows with the space or display property set will automatically be moved back to that space / display if the user moves the window to another space / display, regardless of the method used. This is working as intended and will not be changed.

**Comment by koekeishiya** [comment]
Pattern matching now uses `std::regex` and `std::regex_match`.
For regular expression syntax, refer to ECMAscript specification or POSIX for detailed information.

**Comment by koekeishiya** [comment]
The following rule will automatically add a window to the scratchpad, and either keep it visible or instantly hide it.

`kwmc rule owner="Spotify" properties={scratchpad="visible"}`
`kwmc rule owner="Spotify" properties={scratchpad="hidden"}`

**Comment by koekeishiya** [comment]
`kwmc config add-role ..` has been replaced by:
`kwmc owner="" properties={role=""}`

e.g:
`kwmc owner="Emacs" properties={role="AXStandardWindow"}`

**Comment by mickaelperrin** [comment]
Ability to set a floating window always on top.

**Comment by dominiklohmann** [comment]
> Ability to set a floating window always on top.

See #312.

**Comment by koekeishiya** [comment]
Will close this for now. The position / size thingy is questionable, will figure out what to do about that at some point.

**Comment by mickaelperrin** [comment]
Now that we have `config display 0 float-dim`, any chance to get rules to set adequate windows sizes ? even centered would be a nice have !

**Comment by koekeishiya** [comment]
> Now that we have config display 0 float-dim, any chance to get rules to set adequate windows sizes ? even centered would be a nice have ! 

You can create new issue for this, so that it is easier to track.
